### ISSUE TITLE
`functools.wraps` type inference broken when inferring return type of wrapper

### ISSUE THREAD

**Issue by mezuzza** [microsoft/pyright#6080]
**Describe the bug**
`functools.wraps` infers the wrong return type when wrapper type is not explicitly specified.

**Code or Screenshots**
```
from typing import Callable, Awaitable, ParamSpec, TypeVar
from functools import wraps

R = TypeVar("R")
P = ParamSpec("P")

def bind(f: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]:
  @wraps(f)
  async def wrapper(*args: P.args, **kwargs: P.kwargs): # -> R: Adding an explicit return type here fixes the problem.
    return await f(*args, **kwargs)

  return wrapper # Type Error
```
![image](https://github.com/microsoft/pyright/assets/1093430/f922ab92-c0ee-4874-80b9-7b81aa2cabaa)

The issue is that wrapper is inferred to be `_Wrapped[P@bind, Awaitable[R@bind], P@bind.args, R@bind]`. It should infer that the last type parameter is `Awaitable` in some form (though as I understand it, the `Awaitable` typeshed class itself is somehow not ideal). If you add the return type explicitly, pyright infers `Infers _Wrapped[P@bind, Awaitable[R@bind], P@bind.args, Coroutine[Any, Any, R@bind]]` which is correct. Somehow, when you rely on type inference alone for the return type, the Coroutine part is dropped.

**VS Code extension or command-line**
pyright - 1.1.329 via command-line
python - 3.10.12

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
